engine/apps/api/permissions/__init__.py

```diff
@@ -1,5 +1,196 @@
-from .actions import ActionPermission  # noqa: F401
-from .constants import ALL_BASE_ACTIONS, MODIFY_ACTIONS, READ_ACTIONS  # noqa: F401
-from .methods import MethodPermission  # noqa: F401
-from .owner import IsOwner, IsOwnerOrAdmin, IsOwnerOrAdminOrEditor  # noqa: F401
-from .roles import AnyRole, IsAdmin, IsAdminOrEditor, IsEditor, IsStaff, IsViewer  # noqa: F401
+import enum
+import typing
+
+from rest_framework import permissions
+from rest_framework.authentication import BasicAuthentication, SessionAuthentication
+from rest_framework.request import Request
+from rest_framework.views import APIView
+from rest_framework.viewsets import ViewSet, ViewSetMixin
+
+from common.utils import getattrd
+
+ACTION_PREFIX = "grafana-oncall-app"
+RBAC_PERMISSIONS_ATTR = "rbac_permissions"
+RBAC_OBJECT_PERMISSIONS_ATTR = "rbac_object_permissions"
+
+ViewSetOrAPIView = typing.Union[ViewSet, APIView]
+
+
+class GrafanaPermission(typing.TypedDict):
+    action: str
+
+
+class Resources(enum.Enum):
+    ALERT_GROUPS = "alert-groups"
+    INTEGRATIONS = "integrations"
+    ESCALATION_CHAINS = "escalation-chains"
+    SCHEDULES = "schedules"
+    CHATOPS = "chatops"
+    OUTGOING_WEBHOOKS = "outgoing-webhooks"
+    MAINTENANCE = "maintenance"
+    API_KEYS = "api-keys"
+
+    NOTIFICATION_SETTINGS = "notification-settings"
+    USER_SETTINGS = "user-settings"
+    OTHER_SETTINGS = "other-settings"
+
+    _IMPOSSIBLE_TESTING = "impossible-testing"
+
+
+class Actions(enum.Enum):
+    READ = "read"
+    WRITE = "write"
+    ADMIN = "admin"
+
+    _IMPOSSIBLE_TESTING = "impossible-testing"
+
+
+def _generate_permission_string(resource: Resources, action: Actions, ignore_prefix=False) -> str:
+    with_prefix = f"{ACTION_PREFIX}."
+    return f"{'' if ignore_prefix else with_prefix}{resource.value}:{action.value}"
+
+
+def _get_view_action(request: Request, view: ViewSetOrAPIView) -> str:
+    """
+    For right now this needs to support being used in both a ViewSet as well as APIView, we use both interchangably
+
+    Note: `request.method` is returned uppercase
+    """
+    return view.action if isinstance(view, ViewSetMixin) else request.method.lower()
+
+
+class RBACPermission(permissions.BasePermission):
```

Comment:
Can we somehow validate that permissions are specified for all of the actions  for views that use `RBACPermission`? E.g. raising an error when importing the view if some actions miss permissions? Or at least denying access by default if no permissions are specified for an action?

Reply:
good point. For now I've gone with throwing an `AssertionError` if an action is not specified in the `rbac_permissions` dict in [e082c9f](https://github.com/grafana/oncall/pull/634/commits/e082c9ff0c5903066698fd5f99218fc460f2d6b7). If a route should not require permissions then the developer will need to explicitly specify `[]`.